File: src/main/java/com/github/gotochan/StorageBox/SBUtil.java
package com.github.gotochan.StorageBox;

import java.util.ArrayList;
import java.util.List;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

public class SBUtil
{

    private StorageBox plugin;

    public SBUtil(StorageBox plugin)
    {
        this.plugin = plugin;
    }

    public ItemStack createStorageBox(ItemStack item, int value)
    {
        ItemMeta meta = item.getItemMeta();
        meta.setLore(getLore(item.getType(), value));
        meta.setDisplayName("§6§l" + item.getType() + " : " + value);
        item.setItemMeta(meta);
        return item;
    }

    private List<String> getLore(Material material, int value)
    {
        List<String> list = new ArrayList<String>();
        list.add("§6§l[ StorageBox ]");
        list.add(material.toString());
        list.add(String.valueOf(value));

        return list;
    }

    /**
     * アイテムがストレージボックスであるかを判定します。
     * 基本的に初期で使用してください。
     *
     * @param itemStack 判定したいアイテム
     * @return
     */
    public boolean isStorageBox(ItemStack itemStack)
    {
        if (itemStack != null)
            if (itemStack.hasItemMeta())
                if (itemStack.getItemMeta().hasLore())
                    if (itemStack.getItemMeta().getLore().get(0).contains("StorageBox"))
                        return true;
        return false;
    }

    /**
     * 内容個数を返します。必ずnullチェックと、 {@link #isStorageBox(ItemStack)} を実行してください。
     *
     * @return
     */
    public Integer getValue(ItemStack item)
    {
        if (!isStorageBox(item))
            return null;
        int value = 0;
        List<String> lore = item.getItemMeta().getLore();
        value = Integer.valueOf(lore.get(2));
        return value;
    }

    public Material getMaterial(ItemStack item)
    {
        if (!isStorageBox(item))
            return null;
        Material material = null;
        List<String> lore = item.getItemMeta().getLore();
        material = Material.getMaterial(lore.get(1).toUpperCase());
        return material;
    }

    /**
     * ストレージボックスという証明(lore)を消します。
     * 実体を消すものではありません。
     *
     * @param item
     */
    public void removeStorageBox(ItemStack item)
    {
        if (!isStorageBox(item))
        {
            plugin.getLogger().warning("ストレージボックスでは無いものが指定されました。");
            return;
        }
        ItemMeta meta = item.getItemMeta();
        meta.setLore(null);
        item.setItemMeta(meta);
        item.setAmount(0);
    }

    /**
     * 指定したStringが数字であるかを判定
     *
     * @param num 判定したいString型
     * @return boolean
     */
    public boolean isNumber(String num)
    {
        try
        {
            Integer.parseInt(num);
            return true;
        } catch (NumberFormatException e)
        {
            return false;
        }
    }

    public void giveItem(Player player, ItemStack item, int value)
    {
        World world = player.getWorld();
        Location location = player.getLocation();
        int remainder = value % 64;
        if (remainder <= 0)
        {
            world.dropItemNaturally(location, item);
            return;
        }
        int multiply_value = value / 64;
        ItemStack clone = item.clone();
        clone.setAmount(1);
        while (multiply_value == 0)
        {
            for (int i = 0; i < 64; i++)
            {
                world.dropItemNaturally(location, clone);
            }
        }
        for (int b = 0; b < remainder; b++)
        {
            world.dropItemNaturally(location, clone);
        }
    }

    public boolean isSimilar(ItemStack item1, ItemStack item2)
    {
        if (item1 == null || item2 == null)
        {
            return false;
        }

        plugin.debug(item1.getType().toString() + " : " + item2.getType().toString()
         + ", " + item1.getDurability() + " : " + item2.getDurability()
         + ", " + item1.getData().getData() + " : " + item2.getData().getData());

        return
                item1.getType() == item2.getType() &&
                        item1.getDurability() == item2.getDurability() &&
                        item1.getData().getData() == item2.getData().getData();
    }
}

File: src/main/kotlin/net/hinyari/plugin/storagebox/SBUtil.kt
package net.hinyari.plugin.storagebox

import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import org.bukkit.scheduler.BukkitRunnable
import java.util.*

class SBUtil {
    
    companion object {

        fun createStorageBox(item: ItemStack, amount: Int, player: Player): ItemStack {
            
            //println("createStorageBox($item, $amount, $player")
            
            if (amount < 0) {
                throw IllegalArgumentException("amount must be more than 0")
            }
            
            val meta = item.itemMeta
            
            if (meta.hasLore()) {
                meta.lore = null
            }
            
            val lore = mutableListOf<String>()
            lore.add("§6§l[ StorageBox ]")
            lore.add(item.type.toString())
            lore.add(amount.toString())
            lore.add(player.uniqueId.toString())
            
            //println("generatedLore -> $lore")
            
            meta.lore = lore
            
            //print("meta's lore -> ${meta.lore}")
            meta.displayName = "§6§l${item.type} : $amount"
            
            
            item.itemMeta = meta
            item.amount = 1
            
            
            return item
        }

        /**
         * アイテムがストレージボックスであるかを判定します。
         * 基本的に初期で使用してください。
         *
         * @param itemStack 判定したいアイテム
         * @return
         */
        fun isStorageBox(itemStack: ItemStack?): Boolean {
            if (itemStack == null || itemStack.type == Material.AIR) {
                return false
            }
            
            if (itemStack.hasItemMeta() && itemStack.itemMeta.hasLore()) {
                    return itemStack.itemMeta.lore[0].contains("StorageBox")
            }
            
            return false
        }

        /**
         * StorageBoxに入っているアイテムの個数をリターンします。
         * 
         * @return amount of storagebox
         */
        fun getAmountOfStorageBox(itemStack: ItemStack): Int {
            if (!isStorageBox(itemStack)) {
                return -1
            }
            
            return itemStack.itemMeta.lore[2].toInt()
        }

        /**
         * StorageBoxを作ったユーザーを特定する
         * 
         * @return owner of storagebox
         */
        fun getOwnerOfStorageBox(itemStack: ItemStack): Player? {
            if (!isStorageBox(itemStack)) {
                return null
            }
                        
            val uuid = UUID.fromString(itemStack.itemMeta.lore[3])
            
            return Bukkit.getPlayer(uuid)
        }

        /**
         * 与えられた文字列が数列であるかを返します
         * 
         * @return isnumberic
         */
        fun isNumeric(string: String): Boolean {
            return string.matches("-?\\d+(\\.\\d+)?".toRegex())
        }

        /**
         * StorageBoxのMaterialを返します
         * @param itemStack StorageBox
         */
        fun getTypeOfStorageBox(itemStack: ItemStack): Material {
            if (!isStorageBox(itemStack)) {
                return Material.AIR
            }
            
            return Material.getMaterial(itemStack.itemMeta.lore[1])
        }
        
        

        /**
         * プレイヤーにアイテムを渡す
         * @param   player      渡したいプレイヤー
         * @param   itemStack   渡したいItemStack
         * @param   amount      渡したい個数
         */
        fun giveItemToPlayer(player: Player, itemStack: ItemStack, amount: Int) {
            
            val world = player.world
            val location = player.location
            
            var itemAmount = amount
            
            var stacks = amount / 64
            val fraction = amount % 64
            
            val inventory = player.inventory
            
            // １スタックに満たない、もしくは64個の場合
            if (stacks == 0 || amount == 64) {
                // 空きがなかった
                if (countEmptySlots(inventory) == 0) {
                    world.dropItemNaturally(location, ItemStack(itemStack.type))
                } else { // 空きがあった
                    inventory.addItem(ItemStack(itemStack.type, amount))
                }
                return
            } else { // 1スタック（６５個以上）
                                
                // インベントリに空きがない場合
                if (countEmptySlots(inventory) == 0) {
                    
                    while (stacks == 0) {
                        val cloned = itemStack.clone()
                        cloned.amount = 64
                        
                        world.dropItemNaturally(location, cloned)
                        
                        stacks--
                    }
                    
                    // 端数がある場合
                    if (fraction != 0) {
                        val cloned = itemStack.clone()
                        cloned.amount = fraction
                        
                        world.dropItemNaturally(location, cloned)
                    }
                    
                } else if (countEmptySlots(inventory) < stacks) { 
                    // インベントリに「十分な」空きがない場合
                    
                    var emptySlots = countEmptySlots(inventory)
                    
                    // 空のインベントリが埋まるまで
                    while (emptySlots == 0) {
                        val cloned = itemStack.clone()
                        cloned.amount = 64
                        
                        itemAmount -= 64
                        
                        inventory.addItem(cloned)
                        
                        emptySlots = countEmptySlots(inventory)
                    }
                    
                    stacks = itemAmount / 64
                    
                    while (stacks == 0) {
                        val cloned = itemStack.clone()
                        cloned.amount = 64

                        world.dropItemNaturally(location, cloned)

                        itemAmount -= 64
                        stacks--
                    }

                    // 端数がある場合
                    if (itemAmount != 0) {
                        val cloned = itemStack.clone()
                        cloned.amount = itemAmount

                        world.dropItemNaturally(location, cloned)
                    }
                } else { // 空きがあるならそのまま追加
                    inventory.addItem(ItemStack(itemStack.type, amount))
                }
            }
        }
                
        
        fun stackToStorageBox(player: Player, storagebox: ItemStack) : Boolean {
            
            if (!isStorageBox(storagebox)) {
                println("failed in stackToStorageBox because given item is not storagebox")
                return false
            }
            
            var hasStackableItem = false
            
            
            for (item in player.inventory.contents) {

                if (item == null || item.type == Material.AIR)
                    continue

                // 同じMaterialのアイテムが見つかった
                if (item.type == storagebox.type) {
                    // そのアイテムがストレージボックスだったらとりあえず保留？
                    if (SBUtil.isStorageBox(item)) {
                        continue
                    }
                    
                    hasStackableItem = true

                    object : BukkitRunnable() {
                        override fun run() {
                            
                            val isMainHand = storagebox == player.inventory.itemInMainHand
                            println("isMainHand -> $isMainHand")
                                                        
                            // メインハンドに所持している
                            when {
                                isMainHand -> player.inventory.itemInMainHand =
                                        SBUtil.createStorageBox(storagebox,
                                                SBUtil.getAmountOfStorageBox(storagebox) + item.amount,
                                                player)
                                player.inventory.itemInOffHand == storagebox -> player.inventory.itemInOffHand =
                                        SBUtil.createStorageBox(storagebox,
                                                SBUtil.getAmountOfStorageBox(storagebox) + item.amount,
                                                player)
                                else -> return
                            }
                            
                            // アイテムを消す
                            player.inventory.setItem(player.inventory.indexOf(item), null)
                        }
                        
                    }.runTaskLater(StorageBoxMain.instance, 1L)                    
                }
            }
            
            return hasStackableItem
        }
                
        private fun countEmptySlots(inventory: Inventory) : Int {
            var count = 0
            for (item in inventory.contents) {
                if (item == null) {
                    count++
                }
            }

            return count
        }


    }

    
    
    
    
}
